pulp_ostree/app/serializers.py

```diff
@@ -5,41 +5,109 @@
     https://docs.pulpproject.org/pulpcore/plugins/plugin-writer/index.html
 """
 from gettext import gettext as _
+from tarfile import is_tarfile
+
+from rest_framework import serializers
 
 from pulpcore.plugin import serializers as platform
+from pulpcore.plugin.models import Artifact
+from pulpcore.plugin.viewsets import NamedModelViewSet
 
 from . import models
 
 
-# FIXME: SingleArtifactContentSerializer might not be the right choice for you.
-# If your content type has no artifacts per content unit, use "NoArtifactContentSerializer".
-# If your content type has many artifacts per content unit, use "MultipleArtifactContentSerializer"
-# If you want create content through upload, use "SingleArtifactContentUploadSerializer"
-# If you change this, make sure to do so on "fields" below, also.
-# Make sure your choice here matches up with the create() method of your viewset.
-class OstreeContentSerializer(platform.SingleArtifactContentSerializer):
-    """
-    A Serializer for OstreeContent.
-
-    Add serializers for the new fields defined in OstreeContent and
-    add those fields to the Meta class keeping fields from the parent class as well.
-
-    For example::
+class OstreeRepoUploadSerializer(serializers.Serializer):
+    """A Serializer class for uploading tarballs to a Pulp OSTree repository."""
 
-    field1 = serializers.TextField()
-    field2 = serializers.IntegerField()
-    field3 = serializers.CharField()
+    artifact = platform.RelatedField(
+        many=False,
+        lookup_field="pk",
+        view_name="artifacts-detail",
+        queryset=Artifact.objects.all(),
+        help_text=_("Artifact representing an OSTree commit."),
+    )
+    repository_name = serializers.CharField()
+
+    def validate(self, data):
+        """Check if the uploaded artifact is a tarball and if the last repository version exists."""
+        new_data = {}
+        new_data.update(self.initial_data)
+
+        latest_version = new_data["repository"].latest_version()
+        if not latest_version:
+            raise serializers.ValidationError(
+                _(
+                    "The latest repository version of '{}' was not found".format(
+                        new_data["repository"]
+                    )
+                )
+            )
+        new_data["latest_version"] = latest_version
+
+        artifact = NamedModelViewSet.get_resource(new_data["artifact"])
+        if not is_tarfile(artifact.file.path):
+            raise serializers.ValidationError(_("The uploaded artifact is not a tar archive file"))
+        new_data["artifact"] = artifact
+
+        return new_data
+
+
+class OstreeCommitSerializer(platform.SingleArtifactContentSerializer):
```

Comment:
@mdellweg, I am adding the validation to the upload pipeline since this serializer is never used for creating commits. It is used for read-only purposes and I believe this will still stand forever because it does not make any sense to allow users to modify commits through the API directly. What do you think about that?

Reply:
Well, I assume I can add the validation to `finalize_new_version`!

Reply:
As long as we never use the serializer for creation, this should be ok.
`finalize_new_version` is too late. Content will have been created, once you get there.

Reply:
Okay, will perform the checks in pulp_ostree_first_stage.

Reply:
The checks are now performed in `artifact = Artifact.init_and_validate(filepath_copy, expected_digests=expected_digests)` when creating dc artifacts.

Reply:
Argh, I am removing these checks because this would require me to uncompress `.filez` objects (https://ostreedev.github.io/ostree/repo/#content-objects; see the last two sentences). Does this mean that we will not be able to verify the integrity of an OSTree repository? :sweat: 

```
[lmjachky@localhost 85]$ gunzip < be2d86a7bd1ad17621691bc0a4cbffeef6b1a23eb757f1d8e38e1c6795e43f.filez > test.file

gzip: stdin: not in gzip format
```

Reply:
The checksum that comes with a commit is just like a git hash for a commit. You can't use that value to validate the commit. 

A commit object comes with "a reference to a dirtree/dirmeta pair of checksums which describe the root directory of the filesystem": https://ostreedev.github.io/ostree/repo/#commit-objects

We could potentially add a feature in the future that would validate the root directory of the filesystem, however, I don't think it's needed now.

Reply:
So it is not the checksum of the artifact?

Reply:
It is. The names of all files are checksums. But, maybe you were referring to the checksums @dkliban referred to.

Reply:
I thought, e.g. in pulp_file, "digest" contains the same value as "artifact.sha256" and i thought about comparing them for validation. Not about calculating an actual digest again. But if they do not match (by design) there is no point in this check.